File: app/src/main/java/com/wavesplatform/wallet/v1/request/TransferTransactionRequest.java
package com.wavesplatform.wallet.v1.request;

import android.util.Log;

import com.google.common.primitives.Bytes;
import com.google.common.primitives.Longs;
import com.wavesplatform.wallet.v1.crypto.Base58;
import com.wavesplatform.wallet.v1.crypto.CryptoProvider;
import com.wavesplatform.wallet.v1.util.SignUtil;

public class TransferTransactionRequest {
    public static int SignatureLength = 64;
    public static int KeyLength = 32;
    public static int MaxAttachmentSize = 140;
    public static int MinFee = 100000;

    public String assetId;
    public String senderPublicKey;
    public String recipient;
    public long amount;
    public long timestamp;
    public String feeAssetId;
    public long fee;
    public String attachment;
    public String signature;

    public transient final int txType = 4;

    public byte[] toSignBytes() {
        try {
            byte[] timestampBytes = Longs.toByteArray(timestamp);
            byte[] assetIdBytes = SignUtil.arrayOption(assetId);
            byte[] amountBytes = Longs.toByteArray(amount);
            byte[] feeAssetIdBytes = SignUtil.arrayOption(feeAssetId);
            byte[] feeBytes = Longs.toByteArray(fee);

            return Bytes.concat(new byte[]{txType},
                    Base58.decode(senderPublicKey),
                    assetIdBytes,
                    feeAssetIdBytes,
                    timestampBytes,
                    amountBytes,
                    feeBytes,
                    Base58.decode(recipient),
                    SignUtil.arrayWithSize(attachment));
        } catch (Exception e) {
            Log.e("Wallet", "Couldn't create transaction sign", e);
            return new byte[0];
        }
    }

    public void sign(byte[] privateKey) {
        if (signature == null) {
            signature = Base58.encode(CryptoProvider.sign(privateKey, toSignBytes()));
        }
    }
}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/net/model/request/TransferTransactionRequest.kt
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.sdk.net.model.request

import android.util.Log

import com.google.common.primitives.Bytes
import com.google.common.primitives.Longs
import com.wavesplatform.sdk.crypto.Base58
import com.wavesplatform.sdk.crypto.CryptoProvider
import com.wavesplatform.sdk.utils.SignUtil

class TransferTransactionRequest {

    var assetId: String? = null
    var senderPublicKey: String? = null
    var recipient: String? = null
    var amount: Long = 0
    var timestamp: Long = 0
    var feeAssetId: String? = null
    var fee: Long = 0
    var attachment: String? = null
    var signature: String? = null

    @Transient
    val txType = 4

    fun toSignBytes(): ByteArray {
        try {
            val timestampBytes = Longs.toByteArray(timestamp)
            val assetIdBytes = SignUtil.arrayOption(assetId!!)
            val amountBytes = Longs.toByteArray(amount)
            val feeAssetIdBytes = SignUtil.arrayOption(feeAssetId!!)
            val feeBytes = Longs.toByteArray(fee)

            return Bytes.concat(byteArrayOf(txType.toByte()),
                    Base58.decode(senderPublicKey!!),
                    assetIdBytes,
                    feeAssetIdBytes,
                    timestampBytes,
                    amountBytes,
                    feeBytes,
                    Base58.decode(recipient!!),
                    SignUtil.arrayWithSize(attachment))
        } catch (e: Exception) {
            Log.e("Wallet", "Couldn't create transaction sign", e)
            return ByteArray(0)
        }

    }

    fun sign(privateKey: ByteArray) {
        if (signature == null) {
            signature = Base58.encode(CryptoProvider.sign(privateKey, toSignBytes()))
        }
    }

    companion object {
        const val SignatureLength = 64
        const val KeyLength = 32
        const val MaxAttachmentSize = 140
        const val MinFee = 100000
    }
}
